source/features/add-tag-to-commits.tsx

```diff
@@ -0,0 +1,67 @@
+import React from 'dom-chef';
+import select from 'select-dom';
+import features from '../libs/features';
+import {v4} from '../libs/api';
+import {getOwnerAndRepo} from '../libs/utils';
+import {tag} from '../libs/icons';
+
+interface Tag {
+	name: string;
+	commitResourcePath: string;
+	commit: string;
+}
+
+async function getTags(after?: string): Promise<Tag[]> {
+	const {ownerName, repoName} = getOwnerAndRepo();
+	const {repository} = await v4(`{
+		repository(owner: "${ownerName}", name: "${repoName}") {
+			refs(first: 100, refPrefix: "refs/tags/"${after ? `, after:"${after}"` : ''}) {
+				pageInfo {
+					hasNextPage
+					endCursor
+				}
+				edges {
+					node {
+						target {
+							... on Tag {
+								name
+								commitResourcePath
+							}
+						}
+					}
+				}
+			}
+		}
+	}`);
+	let tags: Tag[] = repository.refs.edges.map((edge: any) => edge.node.target).filter((tag: Tag) => tag.name && tag.commitResourcePath).map((tag: Tag) => ({...tag, commit: tag.commitResourcePath.split('/')[4]}));
+	if (repository.refs.pageInfo.hasNextPage) {
+		tags = tags.concat(await getTags(repository.refs.pageInfo.endCursor));
```

Comment:
This is doing exactly something that I didn’t want to do: fetch multiple pages of tags on every Commits page 😕

Reply:
From what I've seen 100 tags (1 request) is enough for most repos, the only encounter I had with more than that was [facebook/react](https://github.com/facebook/react)

Reply:
It's true, but also there are repos like https://github.com/nodejs/node

Are 6 sequential requests acceptable? Are these heavy? How difficult would it be to only query the next page if no matching tags are found on the current one?

Reply:
[nodejs/node](https://github.com/nodejs/node) doesn't tag their releases on master so even if you did, you would still pull all 6 requests each time.

Reply:
> fetch multiple pages of tags on every Commits page 

I assume we can cache the tags? At least the older ones.

> Are these heavy?

@dotconnor ⬆️

Reply:
> I assume we can cache the tags? At least the older ones.

Cache where? Our cache is rather small and if we're gonna store them in a regular `browser.storage.local` we also need a cleaning mechanism so it doesn't stay there forever.

> Are these heavy?

I think the definition of heavy is "what's the [API cost](https://developer.github.com/v4/guides/resource-limitations/#returning-a-calls-rate-limit-status)?" and "how long does it take for an answer?"

If the tags appear 5 seconds after dom ready then I probably won't see them.

Reply:
> I assume we can cache the tags? At least the older ones.

Sure we could cache the tags, but you would still have to make the API calls to validate that the remaining commits don't have tags. So caching based on the commit (key) and its associated tags (value), and only make a network request if a commit isn't found in the cache, would work best.

> Cache where? Our cache is rather small and if we're gonna store them in a regular browser.storage.local we also need a cleaning mechanism so it doesn't stay there forever.

I think a simple LRU implementation on top of `browser.storage.local` with lazy saving/cleaning should suffice. 

> Are these heavy?

Speed? Really depends on Github, from my testing its a little all over the place from ~200ms to sometimes a full second.

Size? A full page of tags (100) is about 4KB

Reply:
> I think a simple LRU implementation on top of `browser.storage.local` with lazy saving/cleaning should suffice.

If you want to implement that it would be great, you can keep it as a separate PR and replace the existing cookie cache.